src/api/db/models/Project.ts

```diff
@@ -0,0 +1,746 @@
+          camConfig.deployments = sortDeps(
+            camConfig!.deployments.filter(
+              (dep) => !idMatch(dep._id!, deploymentId),
+            ) as typeof camConfig.deployments,
+          );
+
+          await project.save();
+          return { project, camConfig };
+        },
+        { retries: 2 },
+      );
+      // TODO: we need to reverse the above operation if reMapImagesToDeps fails!
+      await ProjectModel.reMapImagesToDeps({ projId: project._id, camConfig });
+      return camConfig;
+    } catch (err) {
+      if (err instanceof GraphQLError) throw err;
+      throw new InternalServerError(err as string);
+    }
+  }
+
+  static async createLabel(
+    input: gql.CreateProjectLabelInput,
+    context: Context,
+  ): Promise<HydratedProject['labels'][0]> {
+    try {
+      const project = await this.queryById(context.user['curr_project']!);
+
+      if (
+        project.labels.filter((label) => {
+          return label.name.toLowerCase() === input.name.toLowerCase();
+        }).length
+      )
+        throw new DBValidationError(
+          'A label with that name already exists, avoid creating labels with duplicate names',
+        );
+
+      project.labels.push({
+        name: input.name,
+        color: input.color,
+      });
+
+      await project.save();
+
+      return project.labels.pop()!;
+    } catch (err) {
+      if (err instanceof GraphQLError) throw err;
+      throw new InternalServerError(err as string);
+    }
+  }
+
+  static async deleteLabel(
+    input: gql.DeleteProjectLabelInput,
+    context: Context,
+  ): Promise<GenericResponse> {
+    try {
+      const project = await this.queryById(context.user['curr_project']!);
+
+      const label = project.labels?.find((p) => p._id.toString() === input._id.toString());
+      if (!label) throw new DeleteLabelError('Label not found on project');
+
+      const count = await ImageModel.countImagesByLabel([input._id], context);
+
+      if (count > MAX_LABEL_DELETE) {
+        const msg =
+          `This label is already in extensive use (>${MAX_LABEL_DELETE} images) and cannot be ` +
+          ' automatically deleted. Please contact nathaniel[dot]rindlaub@tnc[dot]org to request that it be manually deleted.';
+        throw new DeleteLabelError(msg);
+      }
+
+      await ImageModel.deleteAnyLabels(
+        {
+          labelId: input._id,
+        },
+        context,
+      );
+
+      project.labels.splice(project.labels.indexOf(label), 1);
+
+      const views = project.views.map((view) => {
+        if (!Array.isArray(view.filters.labels)) return view;
+
+        return {
+          ...view,
+          filters: {
+            ...view.filters,
+            labels: view.filters.labels.filter((label) =>
+              project.labels.some((l) => l._id === label),
+            ),
+          },
+        } as mongoose.Types.Subdocument<mongoose.Types.ObjectId> & ViewSchema;
```

Comment:
@alukach does this mean that if we don't explicitly add an `_id` field to a subdocument, and instead rely on Mongoose to add it for us, mongoose's `InferRawDocType` doesn't know about the `_id`? I was wondering this (question 2.a) [here](https://github.com/tnc-ca-geo/animl-api/issues/221).

Reply:
Also are you sure we're using that `mongoose.Types.Subdocument<>` generic correctly? I'm a little confused why we'd just be passing in the `ObjectId` into it and not the entire `ViewSchema`.

Reply:
To be honest, I'm not totally certain as to why this was necessary when I wrote it.  I assume I was reaching for some things to make TS happy for the next line:

https://github.com/tnc-ca-geo/animl-api/blob/b8024302211ff73c8b8f073c8d36652147634ee6/src/api/db/models/Project.ts#L653

However, removing the forced coercion, we still get the same type:

<img width="691" alt="image" src="https://github.com/tnc-ca-geo/animl-api/assets/897290/e4406371-e0bb-45aa-a053-5479dabed3c2">


To answer your actual question about the generic usage, the first arg to the `mongoose.Types.Subdocument` is actually the `idType`: 

https://github.com/Automattic/mongoose/blob/17334785f52e001610200380f10416806c674910/types/types.d.ts#L86-L97